### ISSUE TITLE
ETIMEDOUT not returned when deadline passes in surveyor socket

### ISSUE THREAD

**Issue by andrewstarks** [nanomsg/nanomsg#194]
According to my understanding of the documentation, when a NN_SURVEYOR socket that is blocked on nn_recv reaches NN_SURVEYOR_DEADLINE, the socket returns with -1 and nn_errono is set to ETIMEDOUT.

(Actually, the documentation says that `receive` returns with ETIMEDOUT. But that doesn't make sense, so let's go with what I'm understanding.)

Instead, `EFSM` is returned. I traced it as much as I could, I can't offer any insight as to why.

The problem is that EFSM is an error that should crash the program, whereas ETIMEDOUT is an expected condition that I'm looking/hoping for.

Let me know if this isn't helpful, but here is a complete example (modified from Tim's tutorial):

EDIT: This was tested on the latest commit from master branch on both OS X and Windows 8 64bit.

```C
#include <assert.h>
#include <libc.h>
#include <stdio.h>
#include <nanomsg/nn.h>
#include <nanomsg/survey.h>

#define SERVER "server"
#define CLIENT "client"
#define DATE   "DATE"

int server (const char *url)
{
  int sock = nn_socket (AF_SP, NN_SURVEYOR);
  assert (sock >= 0);
  assert (nn_bind (sock, url) >= 0);
  sleep(1); // wait for connections
  int sz_d = strlen(DATE) + 1; // '\0' too
  printf ("SERVER: SENDING DATE SURVEY REQUEST\n");
  int bytes = nn_send (sock, DATE, sz_d, 0);
  assert (bytes == sz_d);
  while (1)
    {
      char *buf = NULL;
      int bytes = nn_recv (sock, &buf, NN_MSG, 0);

// This does not work either. It's how the original tutorial from Tim tried it.
//      if (bytes == ETIMEDOUT) {
//        printf("ETIMEDOUT.\n\n");
//        break;
//      }
      if (bytes >= 0)
      {
        printf ("SERVER: RECEIVED \"%s\" SURVEY RESPONSE\n", buf);
        nn_freemsg (buf);
      } else if (nn_errno() == ETIMEDOUT) {
        printf("ETIMEDOUT.\n\n");
        break;
      } else {
        printf("ERROR %s\n", nn_strerror(nn_errno()) );
        break;
      }
    }
  return nn_shutdown (sock, 0);
}

int client (const char *url, const char *name)
{
  int sock = nn_socket (AF_SP, NN_RESPONDENT);
  assert (sock >= 0);
  assert (nn_connect (sock, url) >= 0);
  while (1)
    {
      char *buf = NULL;
      int bytes = nn_recv (sock, &buf, NN_MSG, 0);
      if (bytes >= 0)
        {
          nn_freemsg (buf);
          char *d = "some date";
          int sz_d = strlen(d) + 1; // '\0' too
          printf ("CLIENT (%s): SENDING DATE SURVEY RESPONSE\n", name);
          int bytes = nn_send (sock, d, sz_d, 0);
          assert (bytes == sz_d);
        }
    }
  return nn_shutdown (sock, 0);
}

int main (const int argc, const char **argv)
{
  if (strncmp (SERVER, argv[1], strlen (SERVER)) == 0 && argc >= 2)
    return server (argv[2]);
  else if (strncmp (CLIENT, argv[1], strlen (CLIENT)) == 0 && argc >= 3)
    return client (argv[2], argv[3]);
  else
    {
      fprintf (stderr, "Usage: survey %s|%s <URL> <ARG> ...\n",
               SERVER, CLIENT);
      return 1;
    }
}
```

And for convenience:

```sh
gcc survey.c /usr/local/lib/libnanomsg.a -o survey
./survey server ipc:///tmp/survey.ipc & server=$!
./survey client ipc:///tmp/survey.ipc client0 & client0=$!
./survey client ipc:///tmp/survey.ipc client1 & client1=$!
./survey client ipc:///tmp/survey.ipc client2 & client2=$!
sleep 3
kill $server $client0 $client1 $client2
```

**Comment by daniel-j-h** [comment]
If it "looks like a bug", then at least change the documentation or provide a warning.

Right now the documentation says "ETIMEDOUT will be returned". Which is simply wrong: https://github.com/nanomsg/nanomsg/blob/6a333ae616c9dd1893c2a47e2b5a8b52b433fa77/doc/nn_survey.txt#L23-L27

Even the nanomsg tests check for EFSM: https://github.com/nanomsg/nanomsg/blob/6a333ae616c9dd1893c2a47e2b5a8b52b433fa77/tests/survey.c#L71

I lost a great amount of time by adhering to the specification.

**Comment by rodbzro** [comment]
Hi there.

I've created a fix for this in PR #417. Works for me.
